webui/react/src/shared/utils/routes.test.ts

```diff
@@ -0,0 +1,363 @@
+import React from 'react';
+
+import { paths } from 'routes/utils';
+
+import history from '../routes/history';
+
+import type { AnyMouseEvent } from './routes';
+import * as routes from './routes';
+
+const setup = (url = '/', base = 'http://www.example.com') => {
+  const newUrl = new URL(url, base);
+  Object.defineProperty(
+    window,
+    'location',
+    {
+      value: {
+        hash: newUrl.hash,
+        host: newUrl.host,
+        hostname: newUrl.hostname,
+        href: newUrl.href,
+        origin: newUrl.origin,
+        password: newUrl.password,
+        pathname: newUrl.pathname.toString(),
+        port: newUrl.port,
+        protocol: newUrl.protocol,
+        search: newUrl.search,
+        username: newUrl.username,
+      },
+      writable: true,
+    },
+  );
+  const location: Location = window.location;
+  return { location };
+};
+
+beforeEach(() => {
+  setup();
+});
```

Comment:
**question:** should there be a `afterEach()` that clears the `window.location` property? I'm guessing it normally is `undefined` in this jest test environment, but would it persist outside of `routes.test.ts` (aka other tests?)

Reply:
I'm not sure... `setup()` function create `window.location` in `routes.test.ts`. I thought I don't need `afterEach` because `beforeEach` initialize it every time. Let me investigate.

Reply:
i am 99% sure changes to `window` don't persist.

Reply:
Hm I'll add `afterEach` then